provisioning/device/src/CertificateInstaller.cs

```diff
@@ -1,68 +1,31 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-using Microsoft.Azure.Devices.Shared;
-using System;
-using System.Collections.Generic;
 using System.Security.Cryptography.X509Certificates;
+using Microsoft.Azure.Devices.Shared;
 
 namespace Microsoft.Azure.Devices.Provisioning.Client
 {
     internal static class CertificateInstaller
     {
-        private static readonly HashSet<string> s_installedCertificates = new HashSet<string>();
-        private static readonly object s_lock = new object();
-
-        static CertificateInstaller()
-        {
-            try
-            {
-                using (var store = new X509Store(StoreName.CertificateAuthority, StoreLocation.CurrentUser))
-                {
-                    store.Open(OpenFlags.ReadOnly);
-                    foreach (X509Certificate2 certificate in store.Certificates)
-                    {
-                        s_installedCertificates.Add(certificate.Thumbprint);
-                    }
-                }
-            }
-            catch (Exception ex)
-            {
-                if (Logging.IsEnabled)
-                {
-                    Logging.Error(
-                     null,
-                     $"{nameof(CertificateInstaller)} failed to read store: {ex}.");
-                }
-            }
-        }
-
         public static void EnsureChainIsInstalled(X509Certificate2Collection certificates)
         {
             if (certificates == null)
             {
                 return;
             }
 
-            lock (s_lock)
+            using var store = new X509Store(StoreName.CertificateAuthority, StoreLocation.CurrentUser);
+            store.Open(OpenFlags.ReadWrite);
+
+            foreach (X509Certificate2 certificate in certificates)
             {
-                using (var store = new X509Store(StoreName.CertificateAuthority, StoreLocation.CurrentUser))
+                X509Certificate2Collection results = store.Certificates.Find(X509FindType.FindByThumbprint, certificate.Thumbprint, false);
+                if (results.Count == 0)
                 {
-                    store.Open(OpenFlags.ReadWrite);
-
-                    foreach (X509Certificate2 certificate in certificates)
-                    {
-                        if (!s_installedCertificates.Contains(certificate.Thumbprint))
-                        {
-                            if (Logging.IsEnabled)
-                            {
-                                Logging.Info(null, $"{nameof(CertificateInstaller)} adding {certificate.Thumbprint}");
-                            }
+                    Logging.Info(null, $"{nameof(CertificateInstaller)} adding {certificate.Thumbprint}");
 
-                            store.Add(certificate);
-                            s_installedCertificates.Add(certificate.Thumbprint);
-                        }
-                    }
+                    store.Add(certificate);
```

Comment:
Why are we adding this back to the store? Seems like we're just logging the existence of this.

Reply:
The code is only adding it to the store when it wasn't already in the store (the original code captured all the installed cert thumbprints and used that to identify).

As to why it adds it, I'm not 100% sure. I was going to ask our local security guy (CrisPop) for advice about it.

Reply:
TLS RFC requires that the client sends the client certificate _and_ the chain up to (but not including) the issuer trusted by the server. (In this case the issuer is not necessary a CA root - can be any issuer that the server is aware of.)
The server's responsibility is to take the partial chain offered by the client and form a chain with any of the certificates it trusts. If a valid chain is formed then the client is trusted and the TLS handshake resumes.

Some interesting notes:
1. The certificate trusted by the store doesn't need to be sent by the client. This is an optimization that makes the TLS Client Hello smaller.
2. The client cert (or chain) doesn't need to be trusted by the client.

To summarize: 
* it is mandatory to have the intermediates not known by the server within the Intermediate Store (which is called `StoreName.CertificateAuthority` in .NET, _not_ to be confused with `Root`! (https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.storename?view=net-5.0).
* it is not recommended to use this functionality because of race conditions. Instead, it's best to store _all_ certificates in stores (including the client certificate) or in Hardware Security Modules rather than files (HSM drivers on Windows would expose the certs through the store).